go.mod

```diff
@@ -5,7 +5,9 @@ go 1.20
 require (
 	github.com/beorn7/perks v1.0.1
 	github.com/cespare/xxhash/v2 v2.3.0
+	github.com/google/go-cmp v0.6.0
 	github.com/json-iterator/go v1.1.12
+	github.com/klauspost/compress v1.17.8
```

Comment:
Came here from the release page. Most people don't care about zstd, but now every user will have one more dependency. Would be better to wait for https://github.com/golang/go/issues/62513 and avoid extra dependencies.

Why not let stdlib's http package handle compression negotiation? It's kind of out of scope of this library (from my POV at least).

Reply:
It's fair point we could outsource compression + negotiation to separate libs, but we wanted default client_golang without extra deps to handle that automatically for everyone. That's some choice which was already made before by adding `gzip`.

> Most people don't care about zstd, but now every user will have one more dependency.

Fair point, hopefully we can remove it once zstd is in std. Luckly, this dep is trusted, proven and very minimal https://github.com/klauspost/compress/blob/master/go.mod - we could vendor that code, but we chosen to add one little dep -- we are open on feedback on how blocking this is for our users.

> stdlib's http package handle compression negotiation

Is there std http flow for compression negotiation? I don't think so.

Reply:
> Is there std http flow for compression negotiation? I don't think so.

Yes, it turns out it doesn't handle that (I didn't know). My point is that there are many things that can be improved and done that will improve how this library works (bigger picture). But this doesn't mean they should be part of this library. Imagine every library that provides e.g. http handlers also tried to take care of compression, compression negotiation, doing http/3, etc, etc, etc. Dependency sprawl quickly gets out of hand, things work differently, become incompatible. I think unix way is a better approach. Provide solution for one problem, limit the scope, reduce dependencies. But also make it possible for your consumers to solve those problems somehow (composition!). With other libraries or with stdlib or with custom code.

I wonder if it's possible to has zstd compression without this PR? Can I do it today with a previous version of this library? It seems it should be possible to have an http handler that wraps the handler of this library (or any other one) and implements compression. Perhaps this should be a separate generic library that anyone could use, contribute too. I found [the chi middleware](https://pkg.go.dev/github.com/go-chi/chi/middleware#Compress) that does gzip compression. Sould be possible to have a zstd middleware? And then anyone who wants it can use it for all/any http handlers they have, not just for the metrics endpoint.

Reply:
> we wanted default client_golang without extra deps to handle that automatically for everyone.

May I ask what is the motivation for this? How many users of this library have asked for this feature?

If something can be done (even if easily), doesn't mean it should be done. My gut feeling is that it is cheaper (cpu/ram/network resources) to pass uncompressed metrics to a scraper that is running in the same local network (99% of people do this?).  If things are running on Kubernetes, the scraper is quite likely on the same host (as a DaemonSet)! So it's "free", virtually/almost no network cost. In that case compression is a pure waste of CPU time. Shall we add some code that detects this situation and optimizes the compression away? 😁 Why not, just a little dependency on the Kubernetes client? Let's make the library handle this case too! Ok, I hope you see my point.

p.s. I'm very grateful to people who contribute and maintain great libraries like this one. Please don't take this personally.